pkg/source/source.go

```diff
@@ -68,164 +77,249 @@ var _ Source = &Channel{}
+	// Start calls to no-op.
+	canceled bool
+	// destinations are the destination channels of the added event handlers
+	destinations []chan event.GenericEvent
+
+	ctx    context.Context
+	cancel context.CancelFunc
+	group  *errgroup.Group
 }
 
 func (cs *Channel) String() string {
 	return fmt.Sprintf("channel source: %p", cs)
 }
 
-// Start implements Source and should only be called by the Controller.
-func (cs *Channel) Start(
-	ctx context.Context,
-	handler handler.EventHandler,
-	queue workqueue.RateLimitingInterface,
-	prct ...predicate.Predicate) error {
+// Start is internal and should be called only by the Controller to start a goroutine that enqueues
+// reconcile.Requests. It should NOT block, instead, it should start a goroutine and return immediately.
+// The context passed to Start can be used to cancel the blocking operations in the Start method. To cancel the
+// goroutine/ shutdown the source a user should call Stop.
+// Start can be called multiple times, allowing multiple controllers to share a single source.
+func (cs *Channel) Start(ctx context.Context, handler handler.EventHandler, queue workqueue.RateLimitingInterface,
+	predicates ...predicate.Predicate) error {
 	// Source should have been specified by the user.
 	if cs.Source == nil {
 		return fmt.Errorf("must specify Channel.Source")
 	}
 
-	// set the stop channel to be the context.
-	cs.stop = ctx.Done()
+	cs.mu.Lock()
+	defer cs.mu.Unlock()
+	if cs.canceled {
+		return nil
+	}
 
 	// use default value if DestBufferSize not specified
 	if cs.DestBufferSize == 0 {
 		cs.DestBufferSize = defaultBufferSize
 	}
 
-	dst := make(chan event.GenericEvent, cs.DestBufferSize)
-
-	cs.destLock.Lock()
-	cs.dest = append(cs.dest, dst)
-	cs.destLock.Unlock()
-
-	cs.once.Do(func() {
-		// Distribute GenericEvents to all EventHandler / Queue pairs Watching this source
-		go cs.syncLoop(ctx)
-	})
+	// Create a destination channel for the event handler
+	// and add it to the list of destinations
+	destination := make(chan event.GenericEvent, cs.DestBufferSize)
+	cs.destinations = append(cs.destinations, destination)
+
+	// Initialize the group and event loop if this is the first time.
+	if cs.group == nil {
+		// Create a new context and group for the event loop.
+		// We want to store the cancellation function and the sub-context
+		// returned from errgroup.WithContext so that we can pass it to the handlers.
+		ctx, cancel := context.WithCancel(ctx)
+		cs.cancel = cancel
+		cs.group, cs.ctx = errgroup.WithContext(ctx)
+
+		// Start the main event loop, given that this is the first time
+		// we are adding an event handler.
+		cs.group.Go(cs.loop)
+	}
 
-	go func() {
-		for evt := range dst {
-			shouldHandle := true
-			for _, p := range prct {
-				if !p.Generic(evt) {
-					shouldHandle = false
-					break
+	// Start a goroutine to handle events for the event handler.
+	cs.group.Go(func() error {
+	eventloop:
+		for {
+			select {
+			case <-cs.ctx.Done():
```

Comment:
(I know I just had a comment to change it to this yesterday)

I'm a bit confused about the cancel semantics we want to have. With the current code only the ctx of the first Start() call is actually used. So if Start() is called a second time canceling that context won't do anything.

Let's figure out what we want and then modify the code accordingly.

Following example:
.Start(ctx1, ...)
.Start(ctx2, ...)

What should happen if ctx1 or ctx2 is canceled? Is the idea that each of them cancels the entire source? Or is the idea that they only cancel their own go routine?

I assume the later makes more sense.

In this case we would go back to using `ctx.Done()` here, but I gues in addition we have to check for `cs.ctx.Done()` (aka if the group is canceled)?

But when we do this we have to ensure that canceling ctx1 doesn't cancel the entire group.

I think we have a similar problem when using cs.ctx.Done() in other places, e.g. in `loop()`.

Hope this makes sense :)

Reply:
cc @alvaroaleman @vincepri

Reply:
I think the solution is to only allow each source to be started once.
For Kind and Informer sources this can be added without removing functionality for c/r users. For the Channel source however, we need to add some logic that allows us to create multiple sources from the same channel while receiving each event in each source.

Additionally, I think the ctx passed to Start should be the only way to shutdown the source. A Shutdown function can be added to the source, but this function will just wait for the source to shutdown and will signal  that new calls to Start should not do anything (this is based in the Shutdown function in the DynamicSharedInformerFactory https://github.com/kubernetes/kubernetes/blob/589d6f388680b29f7f7961bda4f4262733b54210/staging/src/k8s.io/client-go/dynamic/dynamicinformer/interface.go#L38-L48)